python/ql/src/experimental/semmle/python/frameworks/Stdlib.qll

```diff
@@ -9,3 +9,127 @@ private import semmle.python.dataflow.new.TaintTracking
 private import experimental.semmle.python.Concepts
 private import semmle.python.ApiGraphs
+
+private module LDAP {
+  private module LDAP2 {
+    private class LDAP2QueryMethods extends string {
+      LDAP2QueryMethods() {
+        this in ["search", "search_s", "search_st", "search_ext", "search_ext_s"]
+      }
+    }
+
+    private class LDAP2Query extends DataFlow::CallCfgNode, LDAPQuery::Range {
+      DataFlow::Node ldapNode;
+      string ldapPart;
+
+      LDAP2Query() {
+        exists(DataFlow::AttrRead searchMethod, DataFlow::CallCfgNode initCall |
+          this.getFunction() = searchMethod and
+          initCall = API::moduleImport("ldap").getMember("initialize").getACall() and
+          initCall = searchMethod.getObject().getALocalSource() and
+          searchMethod.getAttributeName() instanceof LDAP2QueryMethods and
+          (
+            ldapNode = this.getArg(0) and
+            ldapPart = "DN"
+            or
+            (
+              ldapNode = this.getArg(2) or
+              ldapNode = this.getArgByName("filterstr")
+            ) and
+            ldapPart = "search_filter"
+          )
+        )
+      }
+
+      override DataFlow::Node getLDAPNode() { result = ldapNode }
+
+      override string getLDAPPart() { result = ldapPart }
+    }
+
+    private class LDAP2EscapeDN extends DataFlow::CallCfgNode, LDAPEscape::Range {
+      DataFlow::Node escapeNode;
+
+      LDAP2EscapeDN() {
+        this = API::moduleImport("ldap").getMember("dn").getMember("escape_dn_chars").getACall() and
+        escapeNode = this.getArg(0)
+      }
+
+      override DataFlow::Node getEscapeNode() { result = escapeNode }
+    }
+
+    private class LDAP2EscapeFilter extends DataFlow::CallCfgNode, LDAPEscape::Range {
+      DataFlow::Node escapeNode;
+
+      LDAP2EscapeFilter() {
+        this =
+          API::moduleImport("ldap").getMember("filter").getMember("escape_filter_chars").getACall() and
+        escapeNode = this.getArg(0)
+      }
+
+      override DataFlow::Node getEscapeNode() { result = escapeNode }
+    }
+  }
+
+  private module LDAP3 {
+    private class LDAP3QueryMethods extends string {
+      LDAP3QueryMethods() { this in ["search"] }
+    }
+
+    private class LDAP3Query extends DataFlow::CallCfgNode, LDAPQuery::Range {
+      DataFlow::Node ldapNode;
+      string ldapPart;
+
+      LDAP3Query() {
+        exists(DataFlow::AttrRead searchMethod, DataFlow::CallCfgNode connCall |
+          this.getFunction() = searchMethod and
+          connCall = API::moduleImport("ldap3").getMember("Connection").getACall() and
+          connCall = searchMethod.getObject().getALocalSource() and
+          searchMethod.getAttributeName() instanceof LDAP3QueryMethods and
+          (
+            ldapNode = this.getArg(0) and
+            ldapPart = "DN"
+          )
+          or
+          ldapNode = this.getArg(1) and
+          ldapPart = "search_filter"
+        )
```

Comment:
Looks like this approach to set `ldapPart` depending on whether the argument exists or not is not working fine, since I'm getting results for `this.getArg(0)` both `"DN"` and `"search_filter"`. I like this feature for the user to rapidly identify which part is affected directly looking at the query results, but if it doesn't work fine I'll remove it.

I've tried approaches like the following with no results:

```codeql
          exists( |
            ldapNode = this.getArg(0) and
            ldapPart = "DN"
          )
          or
          exists( |
            ldapNode = this.getArg(1) and
            ldapPart = "search_filter"
          )
```